### PR TITLE
Merge changes during the E3 Hamiltonian model development into main

### PR DESCRIPTION
**Pull request by floatingCatty**
The main changes are listed below:

1. add support for element-specific cutoff in the AtomicData class. The max, can be defined as a dict as {A: ra, B:rb}, the neighbour list will only contain bond A-A for ra, B-B for rb, and A-B for (ra+rb)/2

2. add a dataset to load atomic data style in DeePH's format directly, called DeePHDataset. Add E3statistics in DeePHDataset

3. Add support in load density matrix data, formatted as a h5 file named DM.h5. Refactor the data loading that allows the user to decide whether to load hamiltonian/eigenvalue/overlap/density matrix, by tag: get_hamiltonian/get_eigenvalues/get_overlap/get_DM

4. Refactor the data parsing module to align with the modification above, where the user can choose whether to parse hamiltonian/overlap/eigenvalues/DM

5. updating the E3-Equivariant tensor fitting model, including a nonlocal and a local one.

6. Implementing the SO(2) convolution and its edge-wise parameterization in our method.

7. Accelerating E3 Hamiltonian Rotation with recalculated JD matrix

8. Add diagonal mapping in OrbitalMapper Class.

9. Write a prototype to_model_v1 method in nnsk class to transcript the v2 model to v1 json model.

10. Add many functions in loss module, including: a) support the shift of diagonal elements during E3 Hamiltonian fitting 2) develop the per-bond balanced loss which computed the error of each bond separately, and added up with the same weight to avoid the data scarcity and imbalance. 3) Add HamilLossAnalysis Class to analyze the error (per bond, per element, per irreps, overall) between predicted hamiltonain and the reference one

11. Adjust argcheck accordingly.

12. Fix bugs: a) the ref_batch data are using train_loader's batch information b) iterator error in ref_loader.

### CODE REVIEW

File: dptb/nn/Jd.pt
Comment (QG-phy) [inline dptb/nn/Jd.pt:1]: 这个binary文件是需要的吗？

File: dptb/data/build.py [hunk 1]
```diff
@@ -133,7 +134,21 @@ def build_dataset(set_options, common_options):
...
-    if dataset_type == "DefaultDataset":
+    # input in set_option for Default Dataset:
+    # "root": main dir storing all trajectory folders.
+    #         that is, each subfolder of root contains a trajectory.
+    # "prefix": optional, load selected trajectory folders.
+    # "get_Hamiltonian": load the Hamiltonian file to edges of the graph or not.
+    # "get_eigenvalues": load the eigenvalues to the graph or not.
+    # "setinfo": MUST HAVE, the name of the json file used to build dataset.
+    # Example:
+    # "train": {
```
Comment (QG-phy) [inline dptb/data/build.py:145]: 这部分的注释可以删除，因为我已经加进来 docstring

File: dptb/data/dataset/_deeph_dataset.py [hunk 2]
```diff
@@ -9,71 +9,239 @@
...
+        typed_hopping = {}
+        for bt, tp in idp.bond_to_type.items():
+            hopping_tp_mask = hopping_block_mask[typed_dataset["edge_type"].flatten().eq(tp)]
+            hopping_tp = features[typed_dataset["edge_type"].flatten().eq(tp)]
+            filtered_vecs = torch.where(hopping_tp_mask, hopping_tp, torch.tensor(float('nan')))
+            typed_hopping[bt] = filtered_vecs
+
+        sorted_irreps = idp.orbpair_irreps.sort()[0].simplify()
+        n_scalar = sorted_irreps[0].mul if sorted_irreps[0].ir.l == 0 else 0
+
```
Comment (QG-phy) [inline dptb/data/dataset/_deeph_dataset.py:130]: 这些statistics的代码 在deeph_dataset 和我们自己的default_dataset 一样吗？ 如果一样是不是可以复用啊。

File: dptb/data/dataset/_default_dataset.py [hunk 3]
```diff
@@ -62,21 +67,22 @@ def __init__(self,
...
+            self.data["atomic_numbers"] = atomic_numbers.reshape(self.info["nframes"],natoms)
         else:
             raise ValueError("Wrong atomic_number dimensions.")
         
         # load positions, stored as cartesion no matter what provided.
         pos = np.loadtxt(os.path.join(root, "positions.dat"))
-        assert pos.shape[0] == self.info["nframes"] * self.info["natoms"]
-        pos = pos.reshape(self.info["nframes"], self.info["natoms"], 3)
+        assert pos.shape[0] == self.info["nframes"] * natoms
+        pos = pos.reshape(self.info["nframes"], natoms, 3)
```
Comment (QG-phy) [inline dptb/data/dataset/_default_dataset.py:85]: 所以现在我们不需要在info.json 里面指定natoms了？但是还需要指定nframes 对吧？

File: dptb/data/dataset/_default_dataset.py [hunk 4]
```diff
@@ -120,6 +126,38 @@ def __init__(self,
...
+                eigenvalues = np.expand_dims(eigenvalues, axis=0)
+            assert eigenvalues.shape[0] == self.info["nframes"]
+            assert eigenvalues.shape[1] == self.data["kpoints"].shape[1]
+            self.data["eigenvalues"] = eigenvalues            
+        if get_Hamiltonian==True:
+            assert os.path.exists(os.path.join(self.root, "hamiltonians.h5")), "Hamiltonian file not found."
+            self.data["hamiltonian_blocks"] = h5py.File(os.path.join(self.root, "hamiltonians.h5"), "r")
+        if get_overlap==True:
+            assert os.path.exists(os.path.join(self.root, "overlaps.h5")), "Overlap file not found."
+            self.data["overlap_blocks"] = h5py.File(os.path.join(self.root, "overlaps.h5"), "r")
```
Comment (QG-phy) [inline dptb/data/dataset/_default_dataset.py:156]: 这部分代码，和正上面的代码是**重复**的。你仔细看看。

File: dptb/nn/base.py [hunk 5]
```diff
@@ -284,9 +284,14 @@ def __init__(
 
         if config[-1].get('hidden_feature') is None:
             self.out_layer = AtomicLinear(in_features=config[-1]['in_features'], out_features=config[-1]['out_features'], in_field=out_field, out_field=out_field, device=device, dtype=dtype)
+            nn.init.normal_(self.out_layer.linear.weight, mean=0, std=1e-3)
+            nn.init.normal_(self.out_layer.linear.bias, mean=0, std=1e-3)
         else:
             self.out_layer = AtomicMLP(**config[-1],  if_batch_normalized=False, in_field=in_field, out_field=out_field, activation=activation, device=device, dtype=dtype)
+            nn.init.normal_(self.out_layer.out_layer.weight, mean=0, std=1e-3)
+            nn.init.normal_(self.out_layer.out_layer.bias, mean=0, std=1e-3)
         # self.out_norm = nn.LayerNorm(config[-1]['out_features'], elementwise_affine=True)
```
Comment (QG-phy) [inline dptb/nn/base.py:293]: 这些std 不应该是从 common options 传过来的吗？

File: examples/silicon/nnsk/checkpoint/best_nnsk_b2.600_c2.600_w0.300.pth
Comment (QG-phy) [inline examples/silicon/nnsk/checkpoint/best_nnsk_b2.600_c2.600_w0.300.pth:1]: 这些checkpoint 是什么时候的啊？感觉是旧版的。

File: dptb/data/dataset/_default_dataset.py (hunk 4 above)
Comment (QG-phy) [inline dptb/data/dataset/_default_dataset.py:156]: 这部分代码，和正上面的代码是重复的。你仔细看看。

File: dptb/data/dataset/_deeph_dataset.py [hunk 6]
```diff
@@ -9,71 +9,239 @@
...
-    def len(self) -> int:
-        return self.num_examples
\ No newline at end of file
+    def _E3edgespecies_stat(self, typed_dataset, decay):
+        # we get the bond type marked dataset first
+        idp = self.transform
+        typed_dataset = typed_dataset
+
+        idp.get_irreps(no_parity=False)
+        irrep_slices = idp.orbpair_irreps.slices()
```
Comment (QG-phy) [inline dptb/data/dataset/_deeph_dataset.py:117]: 这些statistics的代码 在deeph_dataset 和我们自己的default_dataset 一样吗？ 如果一样是不是可以复用啊。

File: dptb/nn/hamiltonian.py [hunk 7]
```diff
@@ -299,8 +300,10 @@ def forward(self, data: AtomicDataDict.Type) -> AtomicDataDict.Type:
             # when get the angle, the xyz vector should be transformed to yzx.
             angle = xyz_to_angles(data[AtomicDataDict.EDGE_VECTORS_KEY][:,[1,2,0]]) # (tensor(N), tensor(N))
             # The roataion matrix is SO3 rotation, therefore Irreps(l,1), is used here.
-            rot_mat_L = Irrep(int(l1), 1).D_from_angles(angle[0].cpu(), angle[1].cpu(), torch.tensor(0., dtype=self.dtype)).to(self.device) # tensor(N, 2l1+1, 2l1+1)
-            rot_mat_R = Irrep(int(l2), 1).D_from_angles(angle[0].cpu(), angle[1].cpu(), torch.tensor(0., dtype=self.dtype)).to(self.device) # tensor(N, 2l2+1, 2l2+1)
+            rot_mat_L = wigner_D(int(l1), angle[0], angle[1], torch.zeros_like(angle[0]))
+            rot_mat_R = wigner_D(int(l2), angle[0], angle[1], torch.zeros_like(angle[0]))
+            # rot_mat_L = Irrep(int(l1), 1).D_from_angles(angle[0].cpu(), angle[1].cpu(), torch.tensor(0., dtype=self.dtype)).to(self.device) # tensor(N, 2l1+1, 2l1+1)
+            # rot_mat_R = Irrep(int(l2), 1).D_from_angles(angle[0].cpu(), angle[1].cpu(), torch.tensor(0., dtype=self.dtype)).to(self.device) # tensor(N, 2l2+1, 2l2+1)
```
Comment (QG-phy) [inline dptb/nn/hamiltonian.py:307]: 这个地方为什么修改啊？

Comment (QG-phy) [reply dptb/nn/hamiltonian.py:307]: 之前是从角度计算一个D矩阵，Irrep(int(l1), 1).D_from_angles； 现在是保存了D矩阵，直接调用。

File: dptb/nn/Jd.pt
Comment (floatingCatty) [reply dptb/nn/Jd.pt:1]: 需要的，存的Wigner D matrix

File: dptb/data/dataset/_default_dataset.py (hunk 3 above)
Comment (floatingCatty) [reply dptb/data/dataset/_default_dataset.py:85]: 对的

File: examples/silicon/nnsk/checkpoint/best_nnsk_b2.600_c2.600_w0.300.pth
Comment (floatingCatty) [reply examples/silicon/nnsk/checkpoint/best_nnsk_b2.600_c2.600_w0.300.pth:1]: 是，但我不确定silicon案例里哪些还是要的哪些需要删掉

File: dptb/data/dataset/_default_dataset.py (hunk 4 above)
Comment (floatingCatty) [reply dptb/data/dataset/_default_dataset.py]: 已修改

File: dptb/data/dataset/_deeph_dataset.py (hunk 6 above)
Comment (floatingCatty) [reply dptb/data/dataset/_deeph_dataset.py]: 可以，后面提出来

File: dptb/data/build.py [hunk 8]
```diff
@@ -133,7 +134,21 @@ def build_dataset(set_options, common_options):
...
+    #        "type": "DefaultDataset",
+    #        "root": "foo/bar/data_files_here",
+    #        "prefix": "traj",
+    #        "setinfo": "with_pbc.json"
+    #    }
```
[the remark this answers is not in the document]
Comment (floatingCatty) [reply dptb/data/build.py:150]: 已修改
